[Security Solution][Follow-Up] Make process all rule gaps take a list of rule ids

denar50 opened the pull request:
## Summary
This is a follow up of [this PR](https://github.com/elastic/kibana/pull/227231). 
The existing `processAllRuleGaps` logic receives a rule ID, fetches all the gaps for it and runs a processing function. However, the logic to fetch gaps allows a list of rules, therefore we would like to make this function also take a list of rules to reduce the amount of reads to ES. 

Here is a performance comparison between the existing logic and making `processAllRuleGaps`.

### Without improvements (current implementation):
100 rules, 100 gaps => 2.7s (1.232s deleting gaps)

100 rules, (1-100 gaps) => 2.65s - 3.84s

100 rules, 1000 gaps => 6.69s (4.8s deleting gaps)

### Without improvements (no concurrency):
100 rules, 100 gaps => 5.29s (4.2s deleting gaps)

100 rules, 1000 gaps => 13s (12s deleting gaps)

### With the improvements in this PR:
100 rules, 100 gaps => 2.43s (996ms deleting gaps)

100 rules, (1-100 gaps) => 2.16s - 2.80s

100 rules, 1000 gaps => 8.53s (7.079s deleting gaps)



We are also exposing a `ruleId` property on the `Gap` class in order to keep a count of processed gaps per rule inside `processAllRuleGaps` which is important when `maxProcessedGapsPerRule` is defined (Bulk fill gaps use this limit).

## How to test
You can generate 100 rules, each with 100 gaps using [this tool](https://github.com/elastic/security-documents-generator).
Make sure you add a tag to the rules you create by adding a line to the generated JSON for the create rule request [here](https://github.com/elastic/security-documents-generator/blob/main/src/utils/kibana_api.ts#L180): `tags: ['YOUR-TAG-HERE']`

```
yarn start rules --rules 100 -g 100 -c -i"5m"
```

Call the bulk delete rules endpoint with:
```
curl -X PATCH http://localhost:5601/internal/alerting/rules/_bulk_delete \
  -u elastic:changeme \
  -H 'kbn-xsrf: bla' \
  -H 'x-elastic-internal-origin: Kibana' \
  -H 'Content-Type: application/json' \
  -d '{"filter": "alert.attributes.tags: YOUR-TAG-HERE"}'\
  -w "Total time: %{time_total}\n"
```

denar50 committed:
make process all rule gaps take a list of rule ids

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/321936)
* Commit: 62d4190059f502f54bc7dd6a96fe2a72a5b0c0c0

### Failed CI Steps
* [Rule Management - Prebuilt Rules - Security Solution Cypress Tests #2](https://buildkite.com/elastic/kibana-pull-request/builds/321936#019832ad-6bf3-4b92-a90b-f4440b6a13f4)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/321936#019832ad-6bf3-4b92-a90b-f4440b6a13f4) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/321936/jobs/019832ad-6bf3-4b92-a90b-f4440b6a13f4/artifacts/019832eb-1740-498b-a399-258f93cbcb3e) Rule Management - Prebuilt Rules - Security Solution Cypress Tests #<span></span>2 / Detection rules, Prebuilt Rules Installation and Update workflow Installation of prebuilt rules package via Fleet should install prebuilt rules from the Fleet package should install prebuilt rules from the Fleet package

### History
* :broken_heart: [Build #321863](https://buildkite.com/elastic/kibana-pull-request/builds/321863) failed 47f6d9c6811701c0d9f7fc6777b126169e1843c0
* :broken_heart: [Build #321745](https://buildkite.com/elastic/kibana-pull-request/builds/321745) failed 2cda93d364be2c009ce308a3cfac9ef9f91fc3e2

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/321936","number":321936,"commit":"62d4190059f502f54bc7dd6a96fe2a72a5b0c0c0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/321863","number":321863,"commit":"47f6d9c6811701c0d9f7fc6777b126169e1843c0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/321745","number":321745,"commit":"2cda93d364be2c009ce308a3cfac9ef9f91fc3e2"}],"number":321936}
buildkite-pr-comment-->

denar50 committed:
fix unit test

denar50 commented:
/ci

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

denar50 committed:
expose the ruleId field in the gap class

denar50 committed:
improve transformToGap tests

denar50 commented:
/ci

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

denar50 committed:
fix unit tests: add missing ruleId property to gap objects

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

denar50 committed:
fix linting issue

denar50 commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/323822)
* Commit: 8629bf7ba4f25dc9f066b9f2e4f63fc64e901202

### Failed CI Steps
* [Jest Tests #19](https://buildkite.com/elastic/kibana-pull-request/builds/323822#019850f6-85ad-45e8-b683-f7bc1a0887d2)
* [Jest Tests #19](https://buildkite.com/elastic/kibana-pull-request/builds/323822#0198511d-2580-4cb1-a040-20146ed20e13)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323822#019850f6-85ad-45e8-b683-f7bc1a0887d2) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323822/jobs/019850f6-85ad-45e8-b683-f7bc1a0887d2/artifacts/0198511c-d824-4a47-a632-3fcc6494bdfd) Jest Tests #<span></span>19 / findGaps should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323822#0198511d-2580-4cb1-a040-20146ed20e13) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323822/jobs/0198511d-2580-4cb1-a040-20146ed20e13/artifacts/0198512f-cc99-4190-aaf0-2625020058d6) Jest Tests #<span></span>19 / findGaps should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323822#019850f6-85ad-45e8-b683-f7bc1a0887d2) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323822/jobs/019850f6-85ad-45e8-b683-f7bc1a0887d2/artifacts/0198511c-d823-48a7-8a4a-826bf2b7140a) Jest Tests #<span></span>19 / findGapsById should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323822#0198511d-2580-4cb1-a040-20146ed20e13) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323822/jobs/0198511d-2580-4cb1-a040-20146ed20e13/artifacts/0198512f-cc99-41f1-ad44-fcdba06488d6) Jest Tests #<span></span>19 / findGapsById should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323822#019850f6-85ad-45e8-b683-f7bc1a0887d2) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323822/jobs/019850f6-85ad-45e8-b683-f7bc1a0887d2/artifacts/0198511c-d825-4508-ac0a-49f3e18a3d89) Jest Tests #<span></span>19 / findGapsSearchAfter should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323822#0198511d-2580-4cb1-a040-20146ed20e13) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323822/jobs/0198511d-2580-4cb1-a040-20146ed20e13/artifacts/0198512f-cc9a-4b29-9ae9-de35f13a1f3c) Jest Tests #<span></span>19 / findGapsSearchAfter should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323822#019850f6-85ad-45e8-b683-f7bc1a0887d2) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323822/jobs/019850f6-85ad-45e8-b683-f7bc1a0887d2/artifacts/0198511c-d823-463b-9ad4-210cf03fdd9d) Jest Tests #<span></span>19 / mgetGaps should handle multiple gap documents
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323822#0198511d-2580-4cb1-a040-20146ed20e13) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323822/jobs/0198511d-2580-4cb1-a040-20146ed20e13/artifacts/0198512f-cc9b-44cd-ba99-cced357a5efc) Jest Tests #<span></span>19 / mgetGaps should handle multiple gap documents
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323822#019850f6-85ad-45e8-b683-f7bc1a0887d2) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323822/jobs/019850f6-85ad-45e8-b683-f7bc1a0887d2/artifacts/0198511c-d822-43e1-b30d-d9f5941fa612) Jest Tests #<span></span>19 / mgetGaps should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323822#0198511d-2580-4cb1-a040-20146ed20e13) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323822/jobs/0198511d-2580-4cb1-a040-20146ed20e13/artifacts/0198512f-cc9a-4758-a537-fb3e3de21235) Jest Tests #<span></span>19 / mgetGaps should transform response data to Gap objects

### History
* :broken_heart: [Build #323786](https://buildkite.com/elastic/kibana-pull-request/builds/323786) failed d2ac77fe466b44d795458534c303646154a3fa57
* :broken_heart: [Build #323726](https://buildkite.com/elastic/kibana-pull-request/builds/323726) failed 12f267954dec231649d8064ae3bf0f572131871e
* :green_heart: [Build #322219](https://buildkite.com/elastic/kibana-pull-request/builds/322219) succeeded 96c2fd468acfba167ba3f3daf485b4a0fe4cc161
* :broken_heart: [Build #321936](https://buildkite.com/elastic/kibana-pull-request/builds/321936) failed 62d4190059f502f54bc7dd6a96fe2a72a5b0c0c0
* :broken_heart: [Build #321863](https://buildkite.com/elastic/kibana-pull-request/builds/321863) failed 47f6d9c6811701c0d9f7fc6777b126169e1843c0
* :broken_heart: [Build #321745](https://buildkite.com/elastic/kibana-pull-request/builds/321745) failed 2cda93d364be2c009ce308a3cfac9ef9f91fc3e2

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323822","number":323822,"commit":"8629bf7ba4f25dc9f066b9f2e4f63fc64e901202"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323786","number":323786,"commit":"d2ac77fe466b44d795458534c303646154a3fa57"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323726","number":323726,"commit":"12f267954dec231649d8064ae3bf0f572131871e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/322219","number":322219,"commit":"96c2fd468acfba167ba3f3daf485b4a0fe4cc161"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/321936","number":321936,"commit":"62d4190059f502f54bc7dd6a96fe2a72a5b0c0c0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/321863","number":321863,"commit":"47f6d9c6811701c0d9f7fc6777b126169e1843c0"}],"number":323822}
buildkite-pr-comment-->

denar50 committed:
fix unit tests

denar50 commented on x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/soft_delete/soft_delete_gaps.ts:
```diff
@@ -62,6 +62,6 @@ export const softDeleteGaps = async (params: SoftDeleteGapsParams) => {
     }
   } catch (e) {
-    logger.error(`Failed to soft delete gaps for rule ${ruleId}: ${e.message}`);
+    logger.error(`Failed to soft delete gaps for rules ${ruleIds.join(', ')}: ${e.message}`);
```
Should I print the count of rule ids? I am thinking it is not wise to print the whole list if it is too big

denar50 committed:
rename maxFetchedGaps to maxProcessedGapsPerRule and enforce it per rule

denar50 committed:
The docs of ES recommend to always use the PIT returned in the search request because they might be different
https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-open-point-in-time

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#8966](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8966)
[❌] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/esql/trial_license_complete_tier/configs/ess.config.ts: 75/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F228918%2Fhead)

denar50 committed:
fix linting errors

denar50 commented:
/ci

denar50 committed:
fix linting

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/325850)
* Commit: f7d12d3c212c8800c979f96f37255bb5a183650b

### Failed CI Steps
* [Jest Tests #6](https://buildkite.com/elastic/kibana-pull-request/builds/325850#01987423-2ace-4493-8d0b-a18bad542828)
* [Jest Tests #6](https://buildkite.com/elastic/kibana-pull-request/builds/325850#0198744a-471c-4c62-9830-1a9629083dbb)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/325850#01987423-2ace-4493-8d0b-a18bad542828) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/325850/jobs/01987423-2ace-4493-8d0b-a18bad542828/artifacts/01987449-ef7d-41b6-95f3-400aff5a221f) Jest Tests #<span></span>6 / findGaps should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/325850#0198744a-471c-4c62-9830-1a9629083dbb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/325850/jobs/0198744a-471c-4c62-9830-1a9629083dbb/artifacts/0198745c-61a6-474c-a171-1249c203cbd4) Jest Tests #<span></span>6 / findGaps should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/325850#01987423-2ace-4493-8d0b-a18bad542828) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/325850/jobs/01987423-2ace-4493-8d0b-a18bad542828/artifacts/01987449-ef7d-46bc-829c-d36749eb39c7) Jest Tests #<span></span>6 / findGapsById should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/325850#0198744a-471c-4c62-9830-1a9629083dbb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/325850/jobs/0198744a-471c-4c62-9830-1a9629083dbb/artifacts/0198745c-61a5-4aea-8f61-67d9dd9c5b80) Jest Tests #<span></span>6 / findGapsById should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/325850#01987423-2ace-4493-8d0b-a18bad542828) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/325850/jobs/01987423-2ace-4493-8d0b-a18bad542828/artifacts/01987449-ef7e-44d1-8ec1-005cf95d0e24) Jest Tests #<span></span>6 / findGapsSearchAfter should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/325850#0198744a-471c-4c62-9830-1a9629083dbb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/325850/jobs/0198744a-471c-4c62-9830-1a9629083dbb/artifacts/0198745c-61a5-4ebc-ad04-16d3130fbac3) Jest Tests #<span></span>6 / findGapsSearchAfter should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/325850#01987423-2ace-4493-8d0b-a18bad542828) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/325850/jobs/01987423-2ace-4493-8d0b-a18bad542828/artifacts/01987449-ef7e-4e8b-85e5-63e262483e9b) Jest Tests #<span></span>6 / mgetGaps should handle multiple gap documents
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/325850#0198744a-471c-4c62-9830-1a9629083dbb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/325850/jobs/0198744a-471c-4c62-9830-1a9629083dbb/artifacts/0198745c-61a5-463f-bc82-beb4aad18b3c) Jest Tests #<span></span>6 / mgetGaps should handle multiple gap documents
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/325850#01987423-2ace-4493-8d0b-a18bad542828) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/325850/jobs/01987423-2ace-4493-8d0b-a18bad542828/artifacts/01987449-ef7d-4b98-9700-10961edcbfc8) Jest Tests #<span></span>6 / mgetGaps should transform response data to Gap objects
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/325850#0198744a-471c-4c62-9830-1a9629083dbb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/325850/jobs/0198744a-471c-4c62-9830-1a9629083dbb/artifacts/0198745c-61a5-499a-92bf-d3dc4731db06) Jest Tests #<span></span>6 / mgetGaps should transform response data to Gap objects

### History
* :broken_heart: [Build #325625](https://buildkite.com/elastic/kibana-pull-request/builds/325625) failed e4c9cd6d7f106700d5157296a17eae531b282be1
* :broken_heart: [Build #325603](https://buildkite.com/elastic/kibana-pull-request/builds/325603) failed 7c81953750d9fe3e98fc766beb6fca54cb90db87
* :broken_heart: [Build #325566](https://buildkite.com/elastic/kibana-pull-request/builds/325566) failed bdb6df760cd28fcf471ee5113f4b5a3959d30656
* :green_heart: [Build #323907](https://buildkite.com/elastic/kibana-pull-request/builds/323907) succeeded 3d91268279935de1d9449e479dc4734bcc56946d
* :broken_heart: [Build #323822](https://buildkite.com/elastic/kibana-pull-request/builds/323822) failed 8629bf7ba4f25dc9f066b9f2e4f63fc64e901202

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325850","number":325850,"commit":"f7d12d3c212c8800c979f96f37255bb5a183650b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325625","number":325625,"commit":"e4c9cd6d7f106700d5157296a17eae531b282be1"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325603","number":325603,"commit":"7c81953750d9fe3e98fc766beb6fca54cb90db87"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325566","number":325566,"commit":"bdb6df760cd28fcf471ee5113f4b5a3959d30656"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323907","number":323907,"commit":"3d91268279935de1d9449e479dc4734bcc56946d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323822","number":323822,"commit":"8629bf7ba4f25dc9f066b9f2e4f63fc64e901202"}],"number":325850}
buildkite-pr-comment-->

denar50 committed:
fix unit tests

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

denar50 commented:
@pmuellr exactly. All changes are internal. The changes to the `Gap` class are transient, all underlying event objects remain the same. Please do let me know if you would like a walk through of the changes!

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/328329)
* Commit: 35537f2e61a09c03c7af00d9ac0598ad3f428a39

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #327907](https://buildkite.com/elastic/kibana-pull-request/builds/327907) failed 4d2ea8702bb92794db05778e9362acd07e386b7c
* :yellow_heart: [Build #325904](https://buildkite.com/elastic/kibana-pull-request/builds/325904) was flaky 8b173372082670a2588615026f78801bed6b37d7
* :broken_heart: [Build #325850](https://buildkite.com/elastic/kibana-pull-request/builds/325850) failed f7d12d3c212c8800c979f96f37255bb5a183650b
* :broken_heart: [Build #325625](https://buildkite.com/elastic/kibana-pull-request/builds/325625) failed e4c9cd6d7f106700d5157296a17eae531b282be1
* :broken_heart: [Build #325603](https://buildkite.com/elastic/kibana-pull-request/builds/325603) failed 7c81953750d9fe3e98fc766beb6fca54cb90db87

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/328329","number":328329,"commit":"35537f2e61a09c03c7af00d9ac0598ad3f428a39"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327907","number":327907,"commit":"4d2ea8702bb92794db05778e9362acd07e386b7c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325904","number":325904,"commit":"8b173372082670a2588615026f78801bed6b37d7"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325850","number":325850,"commit":"f7d12d3c212c8800c979f96f37255bb5a183650b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325625","number":325625,"commit":"e4c9cd6d7f106700d5157296a17eae531b282be1"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325603","number":325603,"commit":"7c81953750d9fe3e98fc766beb6fca54cb90db87"}],"number":328329}
buildkite-pr-comment-->

denar50 committed:
Update x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/update/update_gaps.ts

Co-authored-by: Vitalii Dmyterko <92328789+vitaliidm@users.noreply.github.com>

denar50 committed:
Update x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/soft_delete/soft_delete_gaps.ts

Co-authored-by: Vitalii Dmyterko <92328789+vitaliidm@users.noreply.github.com>

denar50 committed:
Update x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/process_all_rule_gaps.ts

Co-authored-by: Vitalii Dmyterko <92328789+vitaliidm@users.noreply.github.com>

denar50 commented on x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/process_all_rule_gaps.ts:
```diff
@@ -32,81 +37,129 @@ export const PROCESS_GAPS_DEFAULT_PAGE_SIZE = 500;
+          );
+
+          ruleIdsToProcess = getNextRuleIdsToProcess(
+            processedGapsCountByRuleId,
+            processedGapsByRuleId,
```
`processedGapsCountByRuleId` is the general count of processed gaps by rule id. 
`processedGapsByRuleId` is the count of processed gaps for the current chunk. 
What this function does is to add the counts of `processedGapsByRuleId` into `processedGapsCountByRuleId`.

I will rename those variables for better clarity.

denar50 replied:
Done

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

denar50 committed:
refactor to improve readability

denar50 commented on x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/process_all_rule_gaps.ts:
```diff
@@ -32,81 +37,129 @@ export const PROCESS_GAPS_DEFAULT_PAGE_SIZE = 500;
+  }
+
+  Object.entries(processedGapsByRuleId).forEach(([ruleId, processedCount]) => {
+    processedGapsCountByRuleId[ruleId] += processedCount;
```
Agree. I fixed this by moving it out of this function

denar50 committed:
move magic number to a proper constant inside find_gaps

denar50 commented on x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/find_gaps.ts:
```diff
@@ -78,12 +78,17 @@ export const findGapsSearchAfter = async ({
-  const { ruleId, start, end, perPage, statuses, sortField, sortOrder } = params;
+  const { ruleIds, start, end, perPage, statuses, sortField, sortOrder } = params;
+
+  if (ruleIds.length > 100) {
```
Most, if not all, our bulk actions have a limit of 100 rules at a time. I have moved that number to a constant as suggested

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/process_gaps_batch.ts:
```diff
@@ -15,38 +15,62 @@ interface ProcessGapsBatchParams {
+  if (maxGapsCountToProcess && maxGapsCountToProcess < gapsClampedIntervals.length) {
+    gapsClampedIntervals = gapsClampedIntervals.slice(
+      0,
+      Math.min(maxGapsCountToProcess, gapsClampedIntervals.length)
```
Agree, fixed.

denar50 commented on x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_delete/bulk_delete_rules.test.ts:
```diff
@@ -212,14 +212,11 @@ describe('bulkDelete', () => {
       unsecuredSavedObjectsClient,
```
This is what I initially did and got results similar to yours.
I discussed this with Marshall, and we noticed that the current case is a bit different. Today we have 10 concurrent "lanes" processing a rule each and whenever we are done with a rule in a lane, we start processing the next one immediately. The new approach takes up to 100 rules and processes them in chunks of 10 in 10 "lanes", meaning that for every measurement we make, we will always be measuring the time of the slowest "lane" since these bulk operations have a max rule count of 100. This is why it is hard to see any gain if we do the measuring as it is.
Therefore we decided to remove the concurrency in the measurements, but I agree that it can be a bit misleading. However in all my measurements, all rules had the same amount of gaps, which might not be the general case in the real world where we would greatly benefit from bulk searching gaps for a list of rules.
cc @marshallmain

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/328710)
* Commit: c65f5009d36ea355f10469f9ca1a7e090e3cd286

### Failed CI Steps
* [Jest Tests #7](https://buildkite.com/elastic/kibana-pull-request/builds/328710#01989e3c-db12-4ed2-8424-c3287d433950)
* [Jest Tests #7](https://buildkite.com/elastic/kibana-pull-request/builds/328710#01989e6b-200c-4fdf-907c-0a974ba849fc)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328710#01989e3c-db12-4ed2-8424-c3287d433950) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328710/jobs/01989e3c-db12-4ed2-8424-c3287d433950/artifacts/01989e6b-0463-4b64-b60c-1aa6afcd8d4c) Jest Tests #<span></span>7 / processAllRuleGaps when nextSearchAfter is null should call closePointInTime when it is done
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328710#01989e6b-200c-4fdf-907c-0a974ba849fc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328710/jobs/01989e6b-200c-4fdf-907c-0a974ba849fc/artifacts/01989e7f-02b0-4cb0-92e5-f8be01a9eef2) Jest Tests #<span></span>7 / processAllRuleGaps when nextSearchAfter is null should call closePointInTime when it is done
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328710#01989e3c-db12-4ed2-8424-c3287d433950) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328710/jobs/01989e3c-db12-4ed2-8424-c3287d433950/artifacts/01989e6b-0464-44a9-9abd-ba46c2f07355) Jest Tests #<span></span>7 / processAllRuleGaps when nextSearchAfter is null should stop fetching gaps
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328710#01989e6b-200c-4fdf-907c-0a974ba849fc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328710/jobs/01989e6b-200c-4fdf-907c-0a974ba849fc/artifacts/01989e7f-02b1-4206-8744-82dbf0d57128) Jest Tests #<span></span>7 / processAllRuleGaps when nextSearchAfter is null should stop fetching gaps

### History
* :broken_heart: [Build #328637](https://buildkite.com/elastic/kibana-pull-request/builds/328637) failed 42d970e90ddf00724a473e54c4d4ee56d216e004
* :green_heart: [Build #328329](https://buildkite.com/elastic/kibana-pull-request/builds/328329) succeeded 35537f2e61a09c03c7af00d9ac0598ad3f428a39
* :broken_heart: [Build #327907](https://buildkite.com/elastic/kibana-pull-request/builds/327907) failed 4d2ea8702bb92794db05778e9362acd07e386b7c
* :yellow_heart: [Build #325904](https://buildkite.com/elastic/kibana-pull-request/builds/325904) was flaky 8b173372082670a2588615026f78801bed6b37d7
* :broken_heart: [Build #325850](https://buildkite.com/elastic/kibana-pull-request/builds/325850) failed f7d12d3c212c8800c979f96f37255bb5a183650b

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/328710","number":328710,"commit":"c65f5009d36ea355f10469f9ca1a7e090e3cd286"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/328637","number":328637,"commit":"42d970e90ddf00724a473e54c4d4ee56d216e004"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/328329","number":328329,"commit":"35537f2e61a09c03c7af00d9ac0598ad3f428a39"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327907","number":327907,"commit":"4d2ea8702bb92794db05778e9362acd07e386b7c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325904","number":325904,"commit":"8b173372082670a2588615026f78801bed6b37d7"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325850","number":325850,"commit":"f7d12d3c212c8800c979f96f37255bb5a183650b"}],"number":328710}
buildkite-pr-comment-->

vitaliidm reviewed (approved):
Thanks for working on this PR, @denar50 

Before merging, let's do testing as suggested in https://github.com/elastic/kibana/pull/228918/files#r2269888326 and update PR with its results

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/329965)
* Commit: ad0669232ece1662cae2b70320d52434f68361df

### Failed CI Steps
* [Jest Tests #7](https://buildkite.com/elastic/kibana-pull-request/builds/329965#0198ada2-ac03-4d10-b601-c0da079268d2)
* [Jest Tests #7](https://buildkite.com/elastic/kibana-pull-request/builds/329965#0198adca-9971-45a7-a925-4f8efa2e400b)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/329965#0198ada2-ac03-4d10-b601-c0da079268d2) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/329965/jobs/0198ada2-ac03-4d10-b601-c0da079268d2/artifacts/0198adca-6e10-4c88-aca1-bc4bb6f89ae5) Jest Tests #<span></span>7 / processAllRuleGaps when nextSearchAfter is null should call closePointInTime when it is done
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/329965#0198adca-9971-45a7-a925-4f8efa2e400b) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/329965/jobs/0198adca-9971-45a7-a925-4f8efa2e400b/artifacts/0198adde-501b-40b1-be15-1a3f67e6bf55) Jest Tests #<span></span>7 / processAllRuleGaps when nextSearchAfter is null should call closePointInTime when it is done
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/329965#0198ada2-ac03-4d10-b601-c0da079268d2) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/329965/jobs/0198ada2-ac03-4d10-b601-c0da079268d2/artifacts/0198adca-6e0f-43a6-8177-5b918be5cfb5) Jest Tests #<span></span>7 / processAllRuleGaps when nextSearchAfter is null should stop fetching gaps
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/329965#0198adca-9971-45a7-a925-4f8efa2e400b) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/329965/jobs/0198adca-9971-45a7-a925-4f8efa2e400b/artifacts/0198adde-501b-49a6-b9e6-9f525a608adf) Jest Tests #<span></span>7 / processAllRuleGaps when nextSearchAfter is null should stop fetching gaps

### History
* :broken_heart: [Build #329649](https://buildkite.com/elastic/kibana-pull-request/builds/329649) failed 8ee4c626472f2b564f2d10877ba7c2175c3bf909
* :broken_heart: [Build #329177](https://buildkite.com/elastic/kibana-pull-request/builds/329177) failed 465ebda7ca98eb3efc23df32cd4be2195a434db2
* :broken_heart: [Build #328710](https://buildkite.com/elastic/kibana-pull-request/builds/328710) failed c65f5009d36ea355f10469f9ca1a7e090e3cd286
* :broken_heart: [Build #328637](https://buildkite.com/elastic/kibana-pull-request/builds/328637) failed 42d970e90ddf00724a473e54c4d4ee56d216e004
* :green_heart: [Build #328329](https://buildkite.com/elastic/kibana-pull-request/builds/328329) succeeded 35537f2e61a09c03c7af00d9ac0598ad3f428a39

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329965","number":329965,"commit":"ad0669232ece1662cae2b70320d52434f68361df"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329649","number":329649,"commit":"8ee4c626472f2b564f2d10877ba7c2175c3bf909"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329177","number":329177,"commit":"465ebda7ca98eb3efc23df32cd4be2195a434db2"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/328710","number":328710,"commit":"c65f5009d36ea355f10469f9ca1a7e090e3cd286"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/328637","number":328637,"commit":"42d970e90ddf00724a473e54c4d4ee56d216e004"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/328329","number":328329,"commit":"35537f2e61a09c03c7af00d9ac0598ad3f428a39"}],"number":329965}
buildkite-pr-comment-->

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

denar50 committed:
fix unit test

denar50 committed:
fix type error

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

denar50 committed:
fix linting problem

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/330766)
* Commit: 61ae9a277d686f81939001a557fd63a56f5c6682

### Failed CI Steps
* [FTR Configs #111](https://buildkite.com/elastic/kibana-pull-request/builds/330766#0198c265-dbe3-49ed-a1c9-7233603ff2d9)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/330766#0198c265-dbe3-49ed-a1c9-7233603ff2d9) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/330766/jobs/0198c265-dbe3-49ed-a1c9-7233603ff2d9/artifacts/0198c298-0f53-4e89-afe1-7f33482bd911) FTR Configs #<span></span>111 / EQL execution logic API @<span></span>ess @<span></span>serverless Alert Suppression for EQL rules @<span></span>skipInServerless sequence queries with suppression duration does not suppress alerts outside of duration

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #330690](https://buildkite.com/elastic/kibana-pull-request/builds/330690) failed 05f8075148b4a692df5f4aa90f79257ac24ce069
* :broken_heart: [Build #330420](https://buildkite.com/elastic/kibana-pull-request/builds/330420) failed a8fdd5a86b70854dc60680e9bff0f193b99e6c34
* :broken_heart: [Build #330307](https://buildkite.com/elastic/kibana-pull-request/builds/330307) failed 13f9f67c6f7b19175e71dd3131a456b85e4d533d
* :broken_heart: [Build #329965](https://buildkite.com/elastic/kibana-pull-request/builds/329965) failed ad0669232ece1662cae2b70320d52434f68361df
* :broken_heart: [Build #329649](https://buildkite.com/elastic/kibana-pull-request/builds/329649) failed 8ee4c626472f2b564f2d10877ba7c2175c3bf909

cc @denar50

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330766","number":330766,"commit":"61ae9a277d686f81939001a557fd63a56f5c6682"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330690","number":330690,"commit":"05f8075148b4a692df5f4aa90f79257ac24ce069"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330420","number":330420,"commit":"a8fdd5a86b70854dc60680e9bff0f193b99e6c34"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330307","number":330307,"commit":"13f9f67c6f7b19175e71dd3131a456b85e4d533d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329965","number":329965,"commit":"ad0669232ece1662cae2b70320d52434f68361df"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329649","number":329649,"commit":"8ee4c626472f2b564f2d10877ba7c2175c3bf909"}],"number":330766}
buildkite-pr-comment-->

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

denar50 committed:
Merge branch 'main' into make-process-all-rule-gaps-function-accept-a-list-of-rule-ids

ymao1 reviewed (approved):
Response Ops changes LGTM. Code review only of rules client related changes.
